File: sdks/capacitor/www/PayCat.js
/**
 * PayCat Cordova Plugin
 */

var exec = require('cordova/exec');

var PayCat = {
  /**
   * Configure the SDK
   * @param {Object} options - Configuration options
   * @param {string} options.apiKey - Your PayCat API key
   * @param {string} [options.appUserId] - Optional app user ID
   * @param {boolean} [options.observerMode] - Observer mode (don't finish transactions)
   * @param {boolean} [options.debugLogsEnabled] - Enable debug logs
   * @param {boolean} [options.useSandbox] - Use sandbox environment
   * @param {Function} success - Success callback
   * @param {Function} error - Error callback
   */
  configure: function(options, success, error) {
    exec(success, error, 'PayCat', 'configure', [options]);
  },

  /**
   * Log in a user
   * @param {string} appUserId - The app user ID
   * @param {Function} success - Success callback with SubscriberInfo
   * @param {Function} error - Error callback
   */
  login: function(appUserId, success, error) {
    exec(success, error, 'PayCat', 'login', [appUserId]);
  },

  /**
   * Log out current user
   * @param {Function} success - Success callback
   * @param {Function} error - Error callback
   */
  logout: function(success, error) {
    exec(success, error, 'PayCat', 'logout', []);
  },

  /**
   * Get subscriber info
   * @param {Function} success - Success callback with SubscriberInfo
   * @param {Function} error - Error callback
   */
  getSubscriberInfo: function(success, error) {
    exec(success, error, 'PayCat', 'getSubscriberInfo', []);
  },

  /**
   * Get available offerings
   * @param {Function} success - Success callback with Offerings
   * @param {Function} error - Error callback
   */
  getOfferings: function(success, error) {
    exec(success, error, 'PayCat', 'getOfferings', []);
  },

  /**
   * Purchase a package
   * @param {string} packageIdentifier - The package identifier
   * @param {string} [offeringIdentifier] - Optional offering identifier
   * @param {Function} success - Success callback with PurchaseResult
   * @param {Function} error - Error callback
   */
  purchasePackage: function(packageIdentifier, offeringIdentifier, success, error) {
    if (typeof offeringIdentifier === 'function') {
      error = success;
      success = offeringIdentifier;
      offeringIdentifier = null;
    }
    exec(success, error, 'PayCat', 'purchasePackage', [packageIdentifier, offeringIdentifier]);
  },

  /**
   * Restore purchases
   * @param {Function} success - Success callback with SubscriberInfo
   * @param {Function} error - Error callback
   */
  restorePurchases: function(success, error) {
    exec(success, error, 'PayCat', 'restorePurchases', []);
  },

  /**
   * Check if user has entitlement
   * @param {string} identifier - Entitlement identifier
   * @param {Function} success - Success callback with {isActive, entitlement}
   * @param {Function} error - Error callback
   */
  checkEntitlement: function(identifier, success, error) {
    exec(success, error, 'PayCat', 'checkEntitlement', [identifier]);
  },

  /**
   * Set user attributes
   * @param {Object} attributes - Key-value pairs of attributes
   * @param {Function} success - Success callback
   * @param {Function} error - Error callback
   */
  setAttributes: function(attributes, success, error) {
    exec(success, error, 'PayCat', 'setAttributes', [attributes]);
  },

  /**
   * Set a single attribute
   * @param {string} key - Attribute key
   * @param {string|null} value - Attribute value (null to remove)
   * @param {Function} success - Success callback
   * @param {Function} error - Error callback
   */
  setAttribute: function(key, value, success, error) {
    var attrs = {};
    attrs[key] = value;
    exec(success, error, 'PayCat', 'setAttributes', [attrs]);
  },

  /**
   * Get paywall data
   * @param {Object} [options] - Options
   * @param {string} [options.identifier] - Paywall identifier
   * @param {string} [options.locale] - Locale code
   * @param {Function} success - Success callback with PaywallResult
   * @param {Function} error - Error callback
   */
  getPaywall: function(options, success, error) {
    if (typeof options === 'function') {
      error = success;
      success = options;
      options = {};
    }
    exec(success, error, 'PayCat', 'getPaywall', [options.identifier || null, options.locale || null]);
  },

  /**
   * Present paywall UI
   * @param {Object} [options] - Options
   * @param {string} [options.identifier] - Paywall identifier
   * @param {string} [options.locale] - Locale code
   * @param {Function} success - Success callback with PaywallPresentResult
   * @param {Function} error - Error callback
   */
  presentPaywall: function(options, success, error) {
    if (typeof options === 'function') {
      error = success;
      success = options;
      options = {};
    }
    exec(success, error, 'PayCat', 'presentPaywall', [options.identifier || null, options.locale || null]);
  },

  /**
   * Track a custom event
   * @param {string} eventName - Event name
   * @param {Object} [properties] - Event properties
   * @param {Function} success - Success callback
   * @param {Function} error - Error callback
   */
  trackEvent: function(eventName, properties, success, error) {
    if (typeof properties === 'function') {
      error = success;
      success = properties;
      properties = null;
    }
    exec(success, error, 'PayCat', 'trackEvent', [eventName, properties]);
  },

  /**
   * Add listener for subscription updates
   * @param {Function} callback - Callback function
   */
  addSubscriptionUpdateListener: function(callback) {
    exec(callback, function() {}, 'PayCat', 'addSubscriptionUpdateListener', []);
  },

  /**
   * Add listener for purchase completed
   * @param {Function} callback - Callback function
   */
  addPurchaseCompletedListener: function(callback) {
    exec(callback, function() {}, 'PayCat', 'addPurchaseCompletedListener', []);
  },

  /**
   * Add listener for purchase failed
   * @param {Function} callback - Callback function
   */
  addPurchaseFailedListener: function(callback) {
    exec(callback, function() {}, 'PayCat', 'addPurchaseFailedListener', []);
  },

  /**
   * Remove all listeners
   * @param {Function} success - Success callback
   * @param {Function} error - Error callback
   */
  removeAllListeners: function(success, error) {
    exec(success, error, 'PayCat', 'removeAllListeners', []);
  }
};

module.exports = PayCat;

File: sdks/kmp/src/commonMain/kotlin/dev/paycat/kmp/PayCat.kt
package dev.paycat.kmp

import dev.paycat.kmp.models.*
import dev.paycat.kmp.network.PayCatAPI
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * PayCat SDK - Main entry point
 * Kotlin Multiplatform implementation
 */
class PayCat private constructor() {

    private var api: PayCatAPI? = null
    private var configuration: PayCatConfiguration? = null
    private var currentAppUserId: String? = null

    private val _subscriberInfo = MutableStateFlow<SubscriberInfo?>(null)
    val subscriberInfo: StateFlow<SubscriberInfo?> = _subscriberInfo.asStateFlow()

    private val _offerings = MutableStateFlow<Offerings?>(null)
    val offerings: StateFlow<Offerings?> = _offerings.asStateFlow()

    companion object {
        private var instance: PayCat? = null

        /**
         * Get singleton instance
         */
        fun getInstance(): PayCat {
            if (instance == null) {
                instance = PayCat()
            }
            return instance!!
        }

        /**
         * Shared instance (alias for getInstance)
         */
        val shared: PayCat
            get() = getInstance()
    }

    /**
     * Check if SDK is configured
     */
    val isConfigured: Boolean
        get() = configuration != null

    /**
     * Check if user is logged in
     */
    val isLoggedIn: Boolean
        get() = currentAppUserId != null

    /**
     * Get current app user ID
     */
    val appUserId: String?
        get() = currentAppUserId

    /**
     * Configure the SDK
     */
    fun configure(config: PayCatConfiguration) {
        configuration = config
        api = PayCatAPI(
            apiKey = config.apiKey,
            useSandbox = config.useSandbox,
            debugEnabled = config.debugLogsEnabled
        )

        // Auto-login if appUserId provided
        config.appUserId?.let {
            currentAppUserId = it
        }

        log("PayCat configured")
    }

    /**
     * Log in user
     */
    suspend fun login(appUserId: String): SubscriberInfo {
        ensureConfigured()

        currentAppUserId = appUserId
        val subscriber = api!!.getSubscriber(appUserId)
        _subscriberInfo.value = subscriber

        log("User logged in: $appUserId")
        return subscriber
    }

    /**
     * Log out user
     */
    fun logout() {
        currentAppUserId = null
        _subscriberInfo.value = null
        _offerings.value = null
        log("User logged out")
    }

    /**
     * Get subscriber info
     */
    suspend fun getSubscriberInfo(): SubscriberInfo {
        ensureConfigured()
        ensureLoggedIn()

        val subscriber = api!!.getSubscriber(currentAppUserId!!)
        _subscriberInfo.value = subscriber
        return subscriber
    }

    /**
     * Get offerings
     */
    suspend fun getOfferings(): Offerings {
        ensureConfigured()

        val offerings = api!!.getOfferings()
        _offerings.value = offerings
        return offerings
    }

    /**
     * Check if user has active entitlement
     */
    suspend fun hasEntitlement(identifier: String): Boolean {
        val subscriber = getSubscriberInfo()
        return subscriber.entitlements[identifier]?.isActive == true
    }

    /**
     * Get entitlement details
     */
    suspend fun getEntitlement(identifier: String): Entitlement? {
        val subscriber = getSubscriberInfo()
        return subscriber.entitlements[identifier]
    }

    /**
     * Set user attributes
     */
    suspend fun setAttributes(attributes: Map<String, String?>) {
        ensureConfigured()
        ensureLoggedIn()

        api!!.setAttributes(currentAppUserId!!, attributes)
        log("Attributes set: ${attributes.keys}")
    }

    /**
     * Set single attribute
     */
    suspend fun setAttribute(key: String, value: String?) {
        setAttributes(mapOf(key to value))
    }

    /**
     * Get paywall template
     */
    suspend fun getPaywall(identifier: String? = null, locale: String = "en"): PaywallTemplate {
        ensureConfigured()
        return api!!.getPaywall(identifier ?: "default", locale)
    }

    /**
     * Track custom event
     */
    suspend fun trackEvent(eventName: String, properties: Map<String, Any>? = null) {
        ensureConfigured()
        api!!.trackEvent(currentAppUserId, eventName, properties)
        log("Event tracked: $eventName")
    }

    /**
     * Restore purchases
     * Platform-specific implementations should override this
     */
    suspend fun restorePurchases(): SubscriberInfo {
        return getSubscriberInfo()
    }

    /**
     * Verify iOS receipt
     */
    suspend fun verifyiOSReceipt(transactionId: String): SubscriberInfo {
        ensureConfigured()
        ensureLoggedIn()

        val subscriber = api!!.verifyReceipt(
            appUserId = currentAppUserId!!,
            platform = "ios",
            transactionId = transactionId
        )
        _subscriberInfo.value = subscriber
        return subscriber
    }

    /**
     * Verify Android purchase
     */
    suspend fun verifyAndroidPurchase(
        productId: String,
        purchaseToken: String
    ): SubscriberInfo {
        ensureConfigured()
        ensureLoggedIn()

        val subscriber = api!!.verifyReceipt(
            appUserId = currentAppUserId!!,
            platform = "android",
            productId = productId,
            purchaseToken = purchaseToken
        )
        _subscriberInfo.value = subscriber
        return subscriber
    }

    // Internal helpers

    private fun ensureConfigured() {
        if (!isConfigured) {
            throw PayCatError.NotConfigured()
        }
    }

    private fun ensureLoggedIn() {
        if (!isLoggedIn) {
            throw PayCatError.NotLoggedIn()
        }
    }

    private fun log(message: String) {
        if (configuration?.debugLogsEnabled == true) {
            println("[PayCat] $message")
        }
    }
}

/**
 * Convenience function to get PayCat instance
 */
fun paycat(): PayCat = PayCat.shared
